File: solution/988. Smallest String Starting From Leaf/988_02.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode], s="") -> str:
        if not root:
            return chr(255)
        
        s = chr(root.val + ord('a')) + s
        
        if root.left == root.right:
            return s
        
        return min(self.smallestFromLeaf(root.left, s),
                   self.smallestFromLeaf(root.right, s))
    

File: solution/988. Smallest String Starting From Leaf/988_02.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    string smallestFromLeaf(TreeNode* root, string s = "") {
        if (!root) return "|";
        
        s = string(1, root->val + 'a') + s;
        
        // if (!root->left && !root->right)
        if (root->left == root->right)
            return s;
        
        return min(smallestFromLeaf(root->left, s), smallestFromLeaf(root->right, s));
    }
};
